File: EPI/15.5/swift/15.5_test.swift
import Darwin

extension Node: CustomStringConvertible {
    public var description: String {
        return describe()
    }

    private func describe(indentation: String = "") -> String {
        let childIndentation = indentation + "  "

        let l = left?.describe(indentation: childIndentation) ?? "nil"
        let r = right?.describe(indentation: childIndentation) ?? "nil"

        return "Node(\n" +
            "\(childIndentation)value: \(self.value),\n" +
            "\(childIndentation)left: \(l),\n" +
            "\(childIndentation)right: \(r)\n" +
            "\(indentation))"
    }
}

typealias NodeFn = (Node) -> Node

let tests: [(tree: Node, subTests: [(nodeA: NodeFn, nodeB: NodeFn, lca: NodeFn)])] = [
    (
        tree: Node(
            value: 10,
            left: Node(value: 5),
            right: Node(value: 15)
        ),
        subTests: [
            (
                nodeA: { $0 },
                nodeB: { $0 },
                lca: { $0 }
            ),
            (
                nodeA: { $0.left! },
                nodeB: { $0.left! },
                lca: { $0.left! }
            ),
            (
                nodeA: { $0.left! },
                nodeB: { $0.right! },
                lca: { $0 }
            ),
        ]
    ),
    (
        tree: Node(
            value: 10,
            left: Node(
                value: 5,
                left: Node(value: 4),
                right: Node(value: 6)
            ),
            right: Node(
                value: 15,
                left: Node(value: 12),
                right: Node(
                    value: 17,
                    left: Node(value: 16),
                    right: Node(value: 20)
                )
            )
        ),
        subTests: [
            (
                nodeA: { $0.left!.left! },
                nodeB: { $0.right! },
                lca: { $0 }
            ),
            (
                nodeA: { $0.left!.left! },
                nodeB: { $0.left! },
                lca: { $0.left! }
            ),
            (
                nodeA: { $0.left!.left! },
                nodeB: { $0.left!.right! },
                lca: { $0.left! }
            ),
            (
                nodeA: { $0.right!.right!.left! },
                nodeB: { $0.right!.right!.right! },
                lca: { $0.right!.right! }
            ),
            (
                nodeA: { $0.right!.right!.left! },
                nodeB: { $0.right!.left! },
                lca: { $0.right! }
            ),
            (
                nodeA: { $0.right!.right!.right! },
                nodeB: { $0.left!.right! },
                lca: { $0 }
            ),
        ]
    ),
]

for test in tests {
    for subTest in test.subTests {
        let nodeA = subTest.nodeA(test.tree)
        let nodeB = subTest.nodeB(test.tree)
        let actual = test.tree.lowestCommonAncestor(nodeA, nodeB)
        let expected = subTest.lca(test.tree)
        guard actual === expected else {
            print("For test tree \(test.tree), " +
                "expected lowest common ancestor of \(nodeA.value) and \(nodeB.value) " +
                "to be \(expected.value), but was \(actual.value)")
            exit(1)
        }
    }
}

File: EPI/Binary Search Trees/Compute the LCA in a BST/typescript/15.5_test.ts
import {Node, lca} from './15.5';

declare function require(name: string): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

const find = (root: Node, value: number): Node =>
    value < root.value ? find(root.left!, value) :
    value > root.value ? find(root.right!, value) :
    root;

interface SubTest {
    a: number;
    b: number;
    lca: number;
}

const tests: {root: Node, subTests: SubTest[]}[] = [
    {
        root: {
            value: 10,
            left: {value: 5},
            right: {value: 15}
        },
        subTests: [
            {a: 10, b: 10, lca: 10},
            {a: 5, b: 5, lca: 5},
            {a: 5, b: 15, lca: 10},
        ]
    },
    {
        root: {
            value: 10,
            left: {
                value: 5,
                left: {value: 4},
                right: {value: 6}
            },
            right: {
                value: 15,
                left: {value: 12},
                right: {
                    value: 17,
                    left: {value: 16},
                    right: {value: 20}
                }
            }
        },
        subTests: [
            {a: 4, b: 15, lca: 10},
            {a: 4, b: 5, lca: 5},
            {a: 4, b: 6, lca: 5},
            {a: 16, b: 20, lca: 17},
            {a: 16, b: 12, lca: 15},
            {a: 20, b: 6, lca: 10},
        ]
    }
];

for (const test of tests) {
    for (const subTest of test.subTests) {
        const actual = lca(test.root, find(test.root, subTest.a), find(test.root, subTest.b)).value;
        assert.strictEqual(actual, subTest.lca,
            `For root ${inspect(test.root)}, a ${subTest.a} and b ${subTest.b}, ` +
            `expected lca to be ${subTest.lca}, but was ${actual}`);
    }
}
